bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/Capability.java

```diff
@@ -97,10 +97,13 @@ public class Capability {
     protected void beforeNewData() {
         properties = new HashMap<>(thing.getProperties());
         firstLaunch = properties.isEmpty();
-        if (firstLaunch && !moduleType.isLogical()) {
-            String name = moduleType.apiName.isBlank() ? moduleType.name() : moduleType.apiName;
-            properties.put(PROPERTY_MODEL_ID, name);
-            properties.put(PROPERTY_VENDOR, VENDOR);
+        if (firstLaunch) {
+            properties.put(PROPERTY_THING_TYPE_VERSION, Integer.toString(moduleType.thingTypeVersion));
```

Comment:
If your thing was previously created with UI, the model and vendor properties are already set. That means the thingTypeVersion property will never be set ?

Reply:
Why is `thingTypeVersion` set at all by the binding? It is set on creation of the thing (either each time when using file-based or once when using UI) or set/updated by the `ThingManager`.

Reply:
@J-N-K : this binding is very particular with no XML files to describe thing types. Everything is built dynamically. So `thingTypeVersion` has to be set by the code ?
Or maybe you have not understood at all how is handled this property and how and when to set it !

Reply:
Understood. But if the binding handles all this dynamically, it also adds/removes channels dynamically, correct? Why would it need the `ThingManager` to update managed things when it does that all on its own (which is fine, just a little bit complex).

The update code in the `ThingManager` is for modifying the channels of a things whose thing type is XML defined. The changed thing type definition in the XML is only applied to textual things (because they are essentially re-created when the file is loaded) and to new managed things, but not to "old" managed things. The update instructions allow the `ThingManager` to apply the needed steps to bring the "old" managed things in line with the new definition (so that they are essentially like a "new" thing). The `thingTypeVersion` is used to determine which steps need to be taken and to prevent applying the same instruction twice (because there is no need to add channel `foo` if this has already been done before).

So as long as no update instructions are involved, there is no need to set the `thingTypeVersion`. When there is no XML that defines the thing type, I don't see how update instructions would be needed. Maybe I'm missing something here.

Reply:
> If your thing was previously created with UI, the model and vendor properties are already set. That means the thingTypeVersion property will never be set ?

No, the discovery process does not add any property. This is only done at first start (identified by the absence of any property) of the handler.

Reply:
> @J-N-K : this binding is very particular with no XML files to describe thing types. Everything is built dynamically. So `thingTypeVersion` has to be set by the code ? Or maybe we have not understood at all how is handled this property and how and when to set it !

You're correct, as there is no XML, I had to add this property in module descriptions enum and then add it by binding code.

Reply:
> Understood. But if the binding handles all this dynamically, it also adds/removes channels dynamically, correct? Why would it need the `ThingManager` to update managed things when it does that all on its own (which is fine, just a little bit complex).
> 

Mmm, correct, maybe. Have to think of it twice. I would not want to have to redevelop all this magic you added to code in the binding to support this feature.

Reply:
> If your thing was previously created with UI, the model and vendor properties are already set. That means the thingTypeVersion property will never be set ?

Correct. And if so, no other mean that asking users to delete and recreate their UI defined things.

Reply:
I didn't check your code, but I have also bindings that have a generic thing type (which only holds the thing configuration) and all channels are added/removed dynamically.

In these bindings I have a `checkChannels` method that runs during `initialize` and checks if all expected channels are present (and adds them when missing) or removes those channels that should not be there. After that the thing is `updateThing`ed (if needed). Since thing initialization is only done once, efficiency is not that important here and the check can be as complex as needed.

Reply:
The only difference with you is that I currently only do it once - the first time the handler is initialized (denoted if the property list is empty).

Reply:
IMO it doesn't hurt to do it every time and it removes the need to keeping track of different thing type versions. You just need to make sure to only update the thing if it was modified, otherwise you could end up in a loop.

Reply:
What is the conclusion of this discussion. Is the code OK now ?

Reply:
Code did not change. @J-N-K prefers doing it at each initialization, I do it only at first initialization. It works, is it worth changing ?

Reply:
Ok